src/lcm/LifeCycleStates.cc

```diff
@@ -2496,3 +2673,500 @@ void LifeCycleManager::trigger_resize_failure(int vid)
+
+                vm->get_disks().backup_disk_ids(backups.do_volatile(), backup_disk_ids);
+
+                for (auto id : backup_disk_ids)
+                {
+                    image->add_backup_disk(id);
+                }
+
+                ipool->update(image.get());
+            }
+        }
+
+        /* ------------------------------------------------------------------ */
+        /* Update backup information for increments                           */
+        /* ------------------------------------------------------------------ */
+
+        if (mode == Backups::INCREMENT)
+        {
+            Increment::Type itype;
+
+            if (incremental_id == -1)
+            {
+                backups.incremental_backup_id(image_id);
+                itype = Increment::FULL;
+            }
+            else
+            {
+                image_id = backups.incremental_backup_id();
+                itype = Increment::INCREMENT;
+            }
+
+            if (auto image = ipool->get(image_id))
+            {
+                int rc = image->add_increment(backups.last_backup_id(), real_sz, itype);
+
+                if ( rc == -1)
+                {
+                    error = "cannot add increment to backup image";
+                    goto error_increment_update;
+                }
+
+                backups.last_increment_id(image->last_increment_id());
+
+                increments = image->increments().total();
+                ds_id      = image->get_ds_id();
+
+                ipool->update(image.get());
+            }
+            else
+            {
+                error = "backup image does not exists";
+                goto error_increment_update;
+            }
+        }
+
+        backups.last_backup_clear();
+
+        if (delete_ids.size() > 0)
+        {
+            // FULL & backups > keep_last
+            ostringstream oss;
+
+            oss << "Removing backup snapshots:";
+
+            for (int i : delete_ids)
+            {
+                oss << " " << i;
+            }
+
+            vm->log("LCM", Log::INFO, oss.str());
+        }
+        else if (keep_last > 0 && increments > keep_last)
+        {
+            // INCREMENTAL & increments > keep_last
+            ostringstream oss;
+
+            oss << "Removing " << increments - keep_last << " backup increments";
+
+            vm->log("LCM", Log::INFO, oss.str());
+
+            //Flag the flatten operation to prevent race coditions with backup
+            backups.active_flatten(true);
+        }
+
+        vmpool->update(vm.get());
+
+        vm.reset();
+
+        /* ------------------------------------------------------------------ */
+        /* Add image to the datastore and forget keep_last backups            */
+        /* ------------------------------------------------------------------ */
+        if (mode == Backups::FULL || incremental_id == -1)
+        {
+            if ( auto ds = dspool->get(ds_id) )
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This snippet may contain an accidental use of incorrect operation. Make sure that the operation being performed is in fact what you intended.